batchconsumer/batchermanager.go

```diff
@@ -0,0 +1,208 @@
+		batchInterval: config.BatchInterval,
+
+		batchMsg:      make(chan tagMsgPair),
+		lastIgnored:   make(chan kcl.SequencePair),
+		lastProcessed: make(chan kcl.SequencePair),
+		shutdown:      make(chan chan<- struct{}),
+	}
+
+	bm.startMessageHandler(bm.batchMsg, bm.lastIgnored, bm.lastProcessed, bm.shutdown)
+
+	return bm
+}
+
+func (b *batcherManager) BatchMessage(tag string, msg []byte, pair kcl.SequencePair) {
+	b.batchMsg <- tagMsgPair{tag, msg, pair}
+}
+
+func (b *batcherManager) LatestIgnored(pair kcl.SequencePair) {
+	b.lastIgnored <- pair
+}
+
+func (b *batcherManager) LatestProcessed(pair kcl.SequencePair) {
+	b.lastProcessed <- pair
+}
+
+func (b *batcherManager) Shutdown() <-chan struct{} {
+	done := make(chan struct{})
+	b.shutdown <- done
+
+	return done
+}
+
+func (b *batcherManager) createBatcher() *batcher {
+	return &batcher{
+		flushCount: b.batchCount,
+		flushSize:  b.batchSize,
+	}
+}
+
+func (b *batcherManager) sendBatch(batcher *batcher, tag string) {
+	if len(batcher.Batch) <= 0 {
+		return
+	}
+
+	err := b.sender.SendBatch(batcher.Batch, tag)
+	switch e := err.(type) {
+	case nil: // Do nothing
+	case PartialSendBatchError:
+		b.log.ErrorD("send-batch", kv.M{"msg": e.Error()})
+		for _, line := range e.FailedMessages {
+			b.log.ErrorD("failed-log", kv.M{"log": line})
+		}
+		stats.Counter("batch-log-failures", len(e.FailedMessages))
+	case CatastrophicSendBatchError:
+		b.log.CriticalD("send-batch", kv.M{"msg": e.Error()})
+		os.Exit(1)
+	default:
+		b.log.CriticalD("send-batch", kv.M{"msg": e.Error()})
+		os.Exit(1)
+	}
+
+	batcher.Clear()
+	stats.Counter("batches-sent", 1)
+}
+
+func (b *batcherManager) sendCheckpoint(
+	tag string, lastIgnoredPair kcl.SequencePair, batchers map[string]*batcher,
+) {
+	smallest := lastIgnoredPair
+
+	for name, batcher := range batchers {
+		if tag == name {
+			continue
+		}
+
+		if len(batcher.Batch) <= 0 {
+			continue
+		}
+
+		// Check for empty because it's possible that no messages have been ignored
+		if smallest.IsEmpty() || batcher.SmallestSeq.IsLessThan(smallest) {
+			smallest = batcher.SmallestSeq
+		}
+	}
+
+	if !smallest.IsEmpty() {
+		b.chkpntManager.Checkpoint(smallest)
+	}
+}
+
+// startMessageDistributer starts a go-routine that routes messages to batches.  It's in uses a
+// go routine to avoid racey conditions.
+func (b *batcherManager) startMessageHandler(
+	batchMsg <-chan tagMsgPair, lastIgnored, lastProcessed <-chan kcl.SequencePair,
+	shutdown <-chan chan<- struct{},
+) {
+	go func() {
+		var lastProcessedPair kcl.SequencePair
+		var lastIgnoredPair kcl.SequencePair
+		batchers := map[string]*batcher{}
+
+		for {
+			for tag, batcher := range batchers { // Flush batcher that hasn't been updated recently
+				if b.batchInterval <= time.Now().Sub(batcher.LastUpdated) {
+					b.sendBatch(batcher, tag)
+					b.sendCheckpoint(tag, lastIgnoredPair, batchers)
+					batcher.Clear()
+				}
+			}
+
+			select {
+			case <-time.NewTimer(time.Second).C:
```

Comment:
would it make sense to setup an external time that sends 'flush' messages on the channel every second?